source/slang/slang-ir-variable-scope-correction.cpp

```diff
@@ -0,0 +1,263 @@
+                if (inst->getOp() == kIROp_Var)
+                {
+                    continue;
+                }
+                workList.add(inst);
+            }
+        }
+    }
+
+    auto instAfterParam = funcInst->getFirstBlock()->getFirstChild();
+    while(instAfterParam->getOp() == kIROp_Param)
+    {
+        instAfterParam = instAfterParam->getNextInst();
+    }
+
+    for(auto inst = workList.begin(); inst != workList.end(); inst++)
+    {
+        if (auto loopHeaderList = loopHeaderMap.tryGetValue(getBlock(*inst)))
+        {
+            _processInstruction(dominatorTree, instAfterParam, *inst, *loopHeaderList, workList);
+        }
+    }
+}
+
+// Check if the instruction is used outside of the loop.
+// The loopHeaderList contains all the loop headers where the original instruction is defined.
+// So we if the block of the user instruction is dominated by the break block of the loop header,
+// it means that it was out of the loop, so it's out of the scope of the loop.
+// Note the reason we use the loopHeaderList is because there could be nested loops, so we need to
+// check all the loop headers from inner to outer.
+bool VariableScopeCorrectionContext::_isOutOfScopeUse(IRInst * userInst, IRDominatorTree* domTree, const List<IRLoop*>& loopHeaderList)
+{
+    if (auto block = getBlock(userInst))
+    {
+        // If the use site of this instruction is dominated by the break block, it means that the
+        // instruction is used after the break block, so we need to make that instruction available globally.
+        // By doing so, we record all the users of this instructions.
+        for(auto loopHeader : loopHeaderList)
+        {
+            auto breakBlock = loopHeader->getBreakBlock();
+            if (domTree->dominates(breakBlock, block))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+void VariableScopeCorrectionContext::_processInstruction(IRDominatorTree* dominatorTree, IRInst* instAfterParam,
+        IRInst* originInst, const List<IRLoop*>& loopHeaderList, List<IRInst*>& workList)
+{
+    List<IRUse*> outOfScopeUsers;
+    for (auto use = originInst->firstUse; use; use=use->nextUse)
+    {
+        if(_isOutOfScopeUse(use->getUser(), dominatorTree, loopHeaderList))
+        {
+            outOfScopeUsers.add(use);
+        }
+    }
+
+    if (outOfScopeUsers.getCount() == 0)
+        return;
+
+    if(_isStorableType(originInst->getDataType()))
+    {
+        _processStorableInst(instAfterParam, originInst, outOfScopeUsers);
+    }
+    else
+    {
+        _processUnstorableInst(originInst, outOfScopeUsers);
+        // After processing the user, we need to add operands of the instruction to the worklist
+        // for later processing.
+        for(UInt idx = 0; idx < originInst->getOperandCount(); idx++)
+        {
+            workList.add(originInst->getOperand(idx));
```

Comment:
We need to use a hash set for the workList to make sure the same inst doesn't get inserted twice into the worklist.